### PR TITLE
[WIP] Permissions

### PR DESCRIPTION
**Pull request by rixx**
Super WIP. Open questions:

## 1. how do we know if somebody can see a room or message

Rooms are generally visible and joinable so far, and we want to limit this in some cases. If you want to limit a room eg to admin users, you'd put `{"room.view": ["admin"]}` into the room's `permission_config`.

Messages/Events are visible and sent to everybody within a room, but moderation stops this from working – if we want to retain original messages, that is?

### 2. Can every room that is visible be joined?

Do we need separate `room.join` and `room.view` permissions?

### 3. How do we show "nobody is allowed" default permissions in general?

Some rooms are only visible to people who are personally in it. How do we handle that? Empty list of required traits is currently "everybody is allowed".

### 4. How do we filter for these things?

Currently we have individual checks ("can I view this object"), which  doesn't allow us to filter querysets. We need a way to filter a room list for rooms visible to the current connection, same for chat events.

Hacky, but maybe worth it: have a `is_visible` DB field, filter for that first, fall back on JSON filters second? This would also solve 3.
I kind of want this for legibility reasons, even though we could probably manage the same thing by a pure JSON filter.

### 5. Do we need per-room banning/silencing?

Banning / silencing is currently implemented world-global (WIP, the actions are tbd), which is probably too brutal?

### CODE REVIEW

**Comment by raphaelm** [comment]
> 1. how do we know if somebody can see a room or message
> Rooms are generally visible and joinable so far, and we want to limit this in some cases. If you want to limit a room eg to admin users, you'd put {"room.view": ["admin"]} into the room's permission_config.

Agreed, rooms are easy enough.

> Messages/Events are visible and sent to everybody within a room, but moderation stops this from working – if we want to retain original messages, that is?

I think we do not need to keep original messages, but even if we want, we could easily filter out redacted messages based on the connected user's permissions, both when fetching chat history as well as with real-time updates. We could just filter in ChatModule.publish_event as the simplest solution.

> 2. Can every room that is visible be joined?
> Do we need separate room.join and room.view permissions?

I think every room that is visible can also be joined. In the future, we might want that e.g. a live stream is visible to everyone but the that only to certain users, but I don't think its ever user-friendly to show a room in the room list if you can't click it. That said, we currently do not have a concept of joining a room, that's fully client-side, we just have a concept of joining chat channels.

> 3. How do we show "nobody is allowed" default permissions in general?
> Some rooms are only visible to people who are personally in it. How do we handle that? Empty list of required traits is currently "everybody is allowed".

You mean like private group conversations? That occured to me recently, and I see three ways forward:

* Each user gets their user ID as an additional trait and we allow the permission model to specify ``OR`` in some way for required traits.

* Each room ID is also a valid trait that we add to users when they are invited. We just need to take care that we don't delete this trait in update_users

* We have an entirely additional concept of permissions for this.

I think the second one is my favorite, but not sure.

> 4. How do we filter for these things?
> Currently we have individual checks ("can I view this object"), which doesn't allow us to filter querysets. We need a way to filter a room list for rooms visible to the current connection, same for chat events.

Chat events are straight-forward: When we fetch them, since they are always fetched for a specific room and we know the rules for the room, so we can build a clear queryset, e.g.:

```
if has_permission('chat.view'):
    ChatEvent.objects.filter(
        room=room,
        is_redacted=has_permission('chat.moderate'),
    )
```

The only thing I think this will be a problem for is rooms, since they are the only thing where the listed thing itself carries its own permissions (``room.view``), that is. Since room lists will probably be rather short, we likely could get away with fetching all rooms from the DB and filtering Python-side, but it ain't nice. Luckily, our database is the most powerful alive and supports this type of query:

```
SELECT * FROM room
WHERE
'["trait1", "trait2"]'::jsonb <@ (room.permission_config#>'{"room.view"}');
```

Didn't look up the Django representation of that query, but if it doesn't exist, I'll be delighted to created it :joy: 

> Hacky, but maybe worth it: have a is_visible DB field, filter for that first, fall back on JSON filters second? This would also solve 3.

Not sure I understand it correctly, in what case would ``is_visible`` be false? A room will always be visible to *someone* probably.

> 5. Do we need per-room banning/silencing?
> Banning / silencing is currently implemented world-global (WIP, the actions are tbd), which is probably too brutal?

I think so?

**Comment by raphaelm** [comment]
Sorry, closed by accident

**Comment by rashfael** [comment]
> You mean like private group conversations? That occured to me recently, and I see three ways forward:
> - Each user gets their user ID as an additional trait and we allow the permission model to specify OR in some way for required traits.
> - Each room ID is also a valid trait that we add to users when they are invited. We just need to take care that we don't delete this trait in update_users
> - We have an entirely additional concept of permissions for this.
>
> I think the second one is my favorite, but not sure.

Parameterized traits seem to add a lot of complexity and more potential for hickups.
If we are doing private channels, we also need to differenciate between owners and members, not only from a permission standpoint, we probably want to show that visually.

Another solution would to keep static traits coupled with a role, and manage wolrd:roles <-> user and channelX:roles <-> user (by having either a list of `[{role: 'admin', channel: xyz}]` inside each user or `{roles: {admin: [a, b, c], member: [e, f, g]}}` inside each channel/other thing which needs permissions.
That way we won't have permissions tied directly to users, which is less powerful, but also less foot-gun-y for admins.

**Comment by raphaelm** [comment]
User <> Traits <> Roles <> Permissions feels too complex to me, to many concepts involved, there must be some simpler solution

File: server/venueless/core/models/auth.py [hunk 1]
```diff
@@ -18,4 +26,16 @@ class Meta:
     def serialize_public(self):
         # Important: If this is updated, venueless.core.services.user.get_public_users also needs to be updated!
         # For performance reasons, it does not use this method directly.
-        return {"id": str(self.id), "profile": self.profile}
+        return {
+            "id": str(self.id),
+            "profile": self.profile,
+            "moderation_state": self.moderation_state,
```
Comment (raphaelm) [inline server/venueless/core/models/auth.py:32]: Is this really public? I don't think any user needs to know if an arbitrary other user is silenced, feels like a data leak

**Comment by raphaelm** [comment]
Another approach for private rooms would just be to set room.view to False for everyone and chat.join as well, but declare that people can *still* see the room if they're already a member and then have a magic invite call that adds them. Would require a concept of membership for eg BBB-only rooms, though

**Comment by rashfael** [comment]
I think we got confused generally. That's my understanding of the words:

Permission: Granular static identifier by which we can check if an action is allowed (maybe in context to something). eg `channel:remove`
Traits: List of something? coming from outside, immutable (save for re-login with different token)
Group: A set of permissions which a user can become part of

I am not sure how traits are supposed to work anymore, what with all the possiblity of dynamic rooms / channels.
Either we interpret traits as basically group names or as some complex data structure of permissions and context, because in the end we will need something like "can moderate in channel X"?

**Comment by rashfael** [comment]
What are our complete requirements?

Things we'll certainly need:
- A certain group of users are able to take any action (admins can join any channel).
- Rooms can be private and are only visible for a user defined list of users (except for example admins, which see all).
- Rooms can be public and accessible by all without needing some permission added.
- Only a user defined list of users can take certain actions per channel (user that created a channel can delete it, invited users cannot).
- Translate magic strings in the token to concrete permissions
- Show the "state" of users to certain other users (admin wants to know if a user is banned or a mod).

Things we possibly want:
- Exclude users from public rooms/channels or even just certain actions in public channels.
- Making it easier to manage permissions for admins by grouping permissions and adding users to groups eg "mod for #general". (this can either be made to be the exclusive way to get permissions or complementary)

Did I miss anything? I'd like to avoid any implementation discussion until we got our requirements down, since adding just one requirement could completely break the design (if we don't go full state-of-the-art permission system with all bells and whistles right away).

**Comment by rixx** [comment]
The discussion isn't really about anything included in the PR, so I'll close it to avoid confusion. We can keep the discussion here and I'll open separate PRs with the usable elements.

**Pull request closed without merging**